contracts/mesh-provider/src/ibc.rs

```diff
@@ -213,33 +221,128 @@ pub fn ack_list_validators(
     for val in res.validators {
         VALIDATORS.save(deps.storage, &val, &Validator::new())?;
     }
-    Ok(IbcBasicResponse::new())
+    RETRIES.update(deps.storage, |mut r| -> Result<crate::state::RetryState, StdError> {
+        r.list_validators_reset();
+        Ok(r)
+    })?;
+    Ok(IbcBasicResponse::new()
+        .add_attribute("action", "ack list_validators"))
 }
 
-pub fn fail_list_validators(_deps: DepsMut) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: send another ListValidators message
-    unimplemented!();
+pub fn fail_list_validators(deps: DepsMut, env: Env) -> Result<IbcBasicResponse, ContractError> {
+    // check if we should retry
+    let mut retries = RETRIES.load(deps.storage)?;
+    if !retries.list_validators_should_retry() {
+        retries.list_validators_reset();
+        RETRIES.save(deps.storage, &retries)?;
+        return Ok(IbcBasicResponse::new().add_event(Event::new("list_validators_fail")))
+    }
+    RETRIES.save(deps.storage, &retries)?;
+
+    // retry
+    let channel_id = CHANNEL.load(deps.storage)?;
+    let packet = ProviderMsg::ListValidators {};
+    let msg = IbcMsg::SendPacket {
+        channel_id: channel_id,
+        data: to_binary(&packet)?,
+        timeout: build_timeout(&env),
+    };
+    Ok(IbcBasicResponse::new()
+        .add_event(Event::new("list_validators_retry"))
+        .add_message(msg))
+}
+
+fn ack_stake(deps: DepsMut, staker: String, validator: String, amount: Uint128) -> Result<IbcBasicResponse, ContractError> {
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Ok(IbcBasicResponse::new().add_event(Event::new("ack_stake_zero_amount")));
+    }
+
+    let mut val = VALIDATORS
```

Comment:
This is basically the code from https://github.com/CosmWasm/mesh-security/pull/38/files#diff-ba0374507dc354d977b251871cc89ff4552166c72b7b326edb14a9686ce8d5bdL118-L128 which is good.

I think the below point was not properly resolved (returning an event rather than erroring in ack), but this case should never happen.